message_list_view: Fix hover on new message.

Ddharmani3 opened the pull request:
<!-- Describe your pull request here.-->
The edit button on the message row is not displayed on hover in certain cases, particularly when there is an auto-scroll and the mouse enters a new message. This issue arises because the mouse hovers over a new message immediately after clicking the send button, and that new message is initially a local copy instead of the actual message. The message rerenders to become the actual copy after a brief delay following a successful message sent.

Consequently, when the message is sent after the hover event, the hovered message is still a local copy, and the `locally_echoed` attribute/flag is set to `true` for that message. This causes the `is_message_editable_ignoring_permissions` function to return `false`, resulting in the `is_content_editable` function also returning false. As a result, the `is_content_editable` argument is `false`, and the edit button is not displayed.

when the message is successfully sent and the message is rerendered, the `message_hover` function is called again on the new message. However, the condition `if ($current_message_hover && rows.id($current_message_hover) === id)` returns `false` at the start of the function and prevents the hover buttons from being updated. To fix this, I removed this condition, I don't feel this is the right approach and would like feedback on this.

Additionally, This PR also fixes when the message has hovered and the allowed time to edit the message set by the organization is expired. The message row didn't update to remove the edit button. Fixed this by adding a setInterval in message_hover, when the message is hovered it constantly checks if `is_content_editable` becomes false while the message is still hovered. If so, it recalls the `message_hover` function which updates the buttons.


Fixes: #25312<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
Edit button not shown:
Before:

https://github.com/zulip/zulip/assets/64723994/b672e8c4-95ed-4158-adb6-280707d86bac

After:

https://github.com/zulip/zulip/assets/64723994/22bcee7b-f5fc-436a-9dd3-db0774c48c1f


The edit button should disappear on edit time expiration(shot by making edit time after send equal to 1 min and then recording after ~50sec) :

https://github.com/zulip/zulip/assets/64723994/526eecb7-9bdf-4315-b88f-ee4cd0d628a8




<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
@zulipbot add "buddy review"

Ddharmani3 commented:
@amanagr This PR is ready for review now.

amanagr commented:
LGTM!

Ddharmani3 commented:
> LGTM! Why is this marked as a draft?

Thanks for reviewing this PR. This PR was drafted as I wanted feedback on the approach of removing `if ($current_message_hover && rows.id($current_message_hover) === id)`. Marked this PR as ready for review now as you have approved of changes.

amanagr commented:
Yeah, we can do somethings to optimize it further but not sure if it worth the effort.

Ddharmani3 commented:
@amanagr resolved the conflicts and I updated the second commit to ensure the logic is correct. In the previous version of the PR, removing the 'if' condition resulted in incorrect hover behaviors. I've also updated the commit message to explain the new logic. I believe this is the correct fix for the issue. Can you please review the approach in the second commit to confirm its correctness?

amanagr commented:
LGTM

timabbott commented on web/src/message_list_view.js:
```diff
@@ -84,6 +86,15 @@ export function message_hover($message_row) {
+                clearInterval(update_edit_button_state);
+                message_hover($message_row);
+            }
+        }, 1000);
```
This is super expensive -- an idle Zulip app is going to be checking this every second that you left the mouse open, which can preventing a laptop from fully spinning down its CPU.

What we should instead do is just calculate the time remaining and do a `setTimeout` -- storing the timer object in a variable so we can cancel it in the `unhover` function.

timabbott replied:
Or maybe better would be to decide that this commit's corner case isn't a bug worth fixing directly; it's the kind of bug that a nicer web framework will fix automatically and doesn't really matter in that it only comes up if you leave your mouse over a message at the moment it's edit deadline runs out -- it's fixed if you move the mouse off and on again.

timabbott commented:
I reworked the first commit in this PR to extract a new `message_list_hover` module, which now contains a pretty large chunk of code from `initialize_kitchen_sink_stuff` as well.

There's still a decent chunk of other stuff that needs to be moved to its future home, but at least this way we avoid making `message_list_view` unnecessarily larger.

With those changes, this looks good to me, merged, thanks @Ddharmani3 and @tridentpancake!

timabbott commented:
Oops, accidentally pushed `main` to this PR, which GitHub unfortunately closes without a way to reopen. Updated version (only change is excluding the new file from requiring node coverage): https://github.com/zulip/zulip/pull/26926
